External/HIP/memmove.hip

```diff
@@ -0,0 +1,191 @@
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <vector>
+
+#include "hip/hip_runtime.h"
+
+// Tests for the functional correctness of the lowering of memmove in device
+// code, including moves with overlapping source and destination ranges. Various
+// memmoves are performed on device side and the result of each is compared to
+// the corresponding operation on the host.
+// Currently, only global memory is tested.
+
+#define VERBOSE 0
+
+#define CHKHIP(r)                                                              \
+  if (r != hipSuccess) {                                                       \
+    std::cerr << hipGetErrorString(r) << std::endl;                            \
+    abort();                                                                   \
+  }
+
+using item_type = uint8_t;
+
+// Maximal number of bytes to copy with a memmove call, used to allocate
+// buffers.
+#define MAX_BYTES_PER_THREAD 2048
+
+// Number of threads that move started in parallel.
+#define NUM_MOVE_THREADS (2 * 32)
+
+// Size of blocks in the grid used for move threads. If the number of threads is
+// smaller than this, it is used instead.
+#define BLOCK_SIZE 256
+
+#define ALLOC_SIZE (2 * NUM_MOVE_THREADS * MAX_BYTES_PER_THREAD)
+
+#define TESTED_FUNCTION __builtin_memmove
+
```

Comment:
Can you also make sure  the stack and shared cases are tested

Reply:
Implemented in cd88789.